urldac7fd8c97_vitalymaryin_unsafe_utils_tgz-pJ8-org_vit_unsafeutils_TcpTestJ8/src/org/vit/unsafeutils/tcp/TcpStreamConsumer.java

```diff
@@ -0,0 +1,227 @@
+                } catch (ConnectionPendingException e) {
+                    logger.error(e.getMessage(), e);
+                    break;
+                } catch (IOException e) {
+                    logger.error(e.getMessage(), e);
+                }
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
+                    logger.error(e.getMessage(), e);
+                }
+            }
+            new Thread(this, "TcpStreamConsumer").start();
+        } catch (UnknownHostException e) {
+            logger.error(e.getMessage(), e);
+        } catch (IOException e) {
+            logger.error(e.getMessage(), e);
+        }
+    }
+
+    protected SocketChannel openSocket() throws IOException {
+        if (debugEnabled)logger.debug("Opening socket");
+        SocketChannel socketChannel = SocketChannel.open();
+        socketChannel.configureBlocking(false);
+        socketChannel.socket().setKeepAlive(true);
+        socketChannel.socket().setTcpNoDelay(true);
+        socketChannel.socket().setReceiveBufferSize(so_RCVBUF);
+        if (debugEnabled) {
+            logger.debug("SO_RCVBUF = "+so_RCVBUF);
+            logger.debug("Socket opened");
+        }
+        return socketChannel;
+    }
+
+    @Override
+    public void sendData(UnsafeSerializable data){
+        throw new UnsupportedOperationException();
+    }
+
+    public void requestAsync(UnsafeSerializable data){
+        if (!connected)
+            return;
+        try {
+            pendingData.get(socket).offer(UnsafeSerializer.write(data));
+            if (selectionOpts == SelectionKey.OP_WRITE){
+                socket.register(selector, SelectionKey.OP_WRITE);
+            }
+            if (debugEnabled) logger.debug("notifying selector");
+            selector.wakeup();
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
+        }
+    }
+
+    public void requestSync(UnsafeSerializable data){
+        if (!connected)
+            return;
+        try {
+            pendingData.get(socket).offer(UnsafeSerializer.write(data));
+            if (selectionOpts == SelectionKey.OP_WRITE){
+                socket.register(selector, SelectionKey.OP_WRITE);
+            }
+            selector.wakeup();
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
+        }
+        final String id = data.getObjectId();
+        Object sync = new Object();
+        if (debugEnabled) logger.debug("sent request for:" + id);
+        long start = System.nanoTime();
+        syncRequestQueue.put(id, sync);
+        synchronized (sync){
+            try {
+                sync.wait(syncReqTMOUT);
+                long stop = System.nanoTime();
+                if (debugEnabled && (stop - start) <= syncReqTMOUT*1000000L) logger.debug("reply for:" + id + " came back in " + (stop-start) + "ns");
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)